sympy/integrals/prde.py

```diff
@@ -359,35 +379,381 @@ def prde_no_cancel_b_small(b, Q, n, DE):
+
+    # Solutions of the original equation are
+    #    y = Sum(dj*fj, (j, 1, r)) + Sum(ei*hi, (i, 1, m)),
+    # where  ei == ci  (i = 1, ..., m),  when
+    # A*Matrix([c1, ..., cm]) == 0 and
+    # B*Matrix([c1, ..., cm, d1, ..., dr]) == 0
+
+    # Build combined constraint matrix with m + r + m columns.
+
+    r = len(f)
+    I = eye(m)
+    A = A.row_join(zeros(A.rows, r + m))
+    B = B.row_join(zeros(B.rows, m))
+    C = I.row_join(zeros(m, r)).row_join(-I)
+
+    return f + H, A.col_join(B).col_join(C)
+
+
+def prde_cancel_primitive(b, Q, n, DE):
+    """
+    Pg, 237
+    """
+    with DecrementLevel(DE):
+        ba, bd = frac_in(b, DE.t)
+        Qyn = [frac_in(q.nth(n), DE.t, field=True) for q in Q]
+        # fn = f1n, ..., frn_n
+        # An is a matrix with coefficients in Const(k)
+        # yn = Sum(djn*fjn, (j, 1, rn))
+        fn, An = param_rischDE(ba, bd, Qyn, DE)
+
+    fn = [Poly(fa.as_expr()/fd.as_expr(), DE.t, field=True)
+            for fa, fd in fn]
+
+    rn = len(fn)
+    m = len(An) - rn
+
+    # there would be a need to form a new equation
+    # just like Dy + fy = Sum(ci*qi, (i, 1, m))
+    # the equation is
+    # Dh + bh = Sum(ci*qi, (i, 1, m)) - Sum(djn*D(fjn*t**n) - b*fj*t**n)
+    # the below thing is wrong
+    for N in range(n, -1, -1):
+        s1 = 0
+        for j in range(rn):
+            djn = An[m + j]
+            fjn = fn[j]
+            s1 += djn*(derivation(fjn*DE.t**n, DE) - b*fjn*DE.t**n)
+
+
+def prde_cancel_exp(b, Q, n, DE):
+    """
+    Pg, 237
+    """
+    with DecrementLevel(DE):
+        # equation (7.23)
+        # Note: b != ba/bd
+        ba, bd = frac_in(b + n*derivation(DE.t, DE)/DE.t, DE.t)
+        Qyn = [(Poly(q.nth(n), DE.t), Poly(1, DE.t)) for q in Q]
+        fn, An = param_rischDE(ba, bd, Qyn, DE)
+
+    fn = [Poly(fa.as_expr()/fd.as_expr(), DE.t, field=True)
+            for fa, fd in fn]
+
+
+def param_poly_rischDE(a, b, q, n, DE):
+    """Polynomial solutions of a parametric Risch differential equation.
+
+    Given a derivation D in k[t], a, b in k[t] relatively prime, and q
+    = [q1, ..., qm] in k[t]^m, return h = [h1, ..., hr] in k[t]^r and
+    a matrix A with m + r columns and entries in Const(k) such that
+    a*Dp + b*p = Sum(ci*qi, (i, 1, m)) has a solution p of degree <= n
+    in k[t] with c1, ..., cm in Const(k) if and only if p = Sum(dj*hj,
+    (j, 1, r)) where d1, ..., dr are in Const(k) and (c1, ..., cm,
+    d1, ..., dr) is a solution of Ax == 0.
+    """
+    m = len(q)
+    if n < 0:
+        # Only the trivial zero solution is possible.
+        # Find relations between the qi.
+        if all([qi.is_zero for qi in q]):
+            return [], zeros(1, m)  # No constraints.
+
+        N = max([qi.degree(DE.t) for qi in q])
+        M = Matrix(N + 1, m, lambda i, j: q[j].nth(i))
+        A, _ = constant_system(M, zeros(M.rows, 1), DE)
+
+        return [], A
+
+    if a.is_ground:
+        # Normalization: a = 1.
+        a = a.LC()
```

Comment:
It seems that something goes wrong in the construction of DE before any integration is attempted:
```
f = (log(-1/y)/2 - log(1/y)/2)/y - (log(1 - 1/y)/2 - log(1 + 1/y)/2)/y

In [4]: pdb.run('risch_integrate(f, y)')
> <string>(1)<module>()
(Pdb) s
--Call--
> /home/jks/test/sympy/sympy/integrals/risch.py(1570)risch_integrate()
-> def risch_integrate(f, x, extension=None, handle_first='log',
(Pdb) 
> /home/jks/test/sympy/sympy/integrals/risch.py(1675)risch_integrate()
-> f = S(f)
(Pdb) n
> /home/jks/test/sympy/sympy/integrals/risch.py(1677)risch_integrate()
-> DE = extension or DifferentialExtension(f, x, handle_first=handle_first,
(Pdb) 
> /home/jks/test/sympy/sympy/integrals/risch.py(1678)risch_integrate()
-> dummy=True, rewrite_complex=rewrite_complex)
(Pdb) 
> /home/jks/test/sympy/sympy/integrals/risch.py(1679)risch_integrate()
-> fa, fd = DE.fa, DE.fd
(Pdb) p DE
DifferentialExtension(extension=dict(fa=Poly(-2*y*_t2 + 2*_t1*y + 2*I*pi*y, _t2, domain='EX'), fd=Poly(4*y**2, _t2, domain='ZZ[y]'), D=[Poly(1, y, domain='ZZ'), Poly(-1/y, _t0, domain='ZZ(y)'), Poly(-1/(y**2 + y), _t1, domain='ZZ(y)'), Poly(1/(y**2 - y), _t2, domain='ZZ(y)')]))
```
The domain of `fa` should not be `EX` but rather something like `ZZ(pi, y)`. I'll try to see how that will affect the workings of prde.